observer/probers/tls/tls.go

```diff
@@ -92,102 +84,80 @@ func getOCSPStatus(cert, issuer *x509.Certificate) (int, error) {
-	p.certExpiry.WithLabelValues(p.url).Set(expiryTime.Seconds())
-	p.certExpiry.WithLabelValues(p.url, badOutcomeError{reason}.Error()).Set(float64(reason))
 }
 
-// Probe performs the configured TLS protocol. Return true if both root AND
-// response are the expected values, otherwise false. Export time to cert expiry
-// and reason as Prometheus metrics.
-func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
+func (p TLSProbe) probeUnexpired(timeout time.Duration) (bool, time.Duration) {
 	start := time.Now()
-	conn, secure_err := tls.Dial("tcp", p.url+":443", &tls.Config{})
-	if secure_err != nil {
-		// TLS connection failed, so try insecure connection instead.
-		rootO, rootCN, timeUntilExpiry, err := p.getInsecureCertInfo()
-		// Insecure connection successful, but certificate chain is invalid.
-		if errors.Is(err, fmt.Errorf(reasonToString[issuerVerifyFailed])) {
-			p.exportMetrics(timeUntilExpiry, issuerVerifyFailed)
-			return false, time.Since(start)
-		}
-		// Insecure connection failed.
-		if err != nil {
-			p.exportMetrics(timeUntilExpiry, internalError)
-			return false, time.Since(start)
-		}
-		// Insecure connection successful and chain is valid, so check if
-		// certificate is and should be expired.
-		if p.response == "expired" {
-			certInvalidErr := x509.CertificateInvalidError{}
-			p.exportMetrics(timeUntilExpiry, none)
-			return (errors.As(secure_err, &certInvalidErr) && certInvalidErr.Reason == x509.Expired) && p.isRootExpected(rootO, rootCN), time.Since(start)
-		}
-		// None of the above cases. Secure TLS connection failed and certificate
-		// should not be expired, so return false.
-		p.exportMetrics(timeUntilExpiry, internalError)
+	conn, err := tls.DialWithDialer(&net.Dialer{Timeout: timeout}, "tcp", p.url+":443", &tls.Config{})
+	if err != nil {
+		p.exportMetrics(time.Time{}, internalError)
 		return false, time.Since(start)
 	}
-	// Certificate has been validated. Now check if it's revoked.
 	defer conn.Close()
-	chains := conn.ConnectionState().VerifiedChains
-	leaf, issuer, rootCert := chains[0][0], chains[0][1], chains[0][len(chains[0])-1].Issuer
-	timeUntilExpiry := time.Until(leaf.NotAfter)
-	rootO, rootCN := rootCert.Organization[0], rootCert.CommonName
+	peers := conn.ConnectionState().PeerCertificates
+	root := peers[len(peers)-1].Issuer
+	if root.Organization[0] != p.rootOrg || root.CommonName != p.rootCN {
+		p.exportMetrics(peers[0].NotAfter, none)
+		return false, time.Since(start)
+	}
 
-	// Check OCSP to see if the certificate is valid, revoked or unknown.
-	ocspStatus, err := getOCSPStatus(leaf, issuer)
+	ocspStatus, err := getOCSP(peers[0], peers[1])
 	if err != nil {
-		p.exportMetrics(timeUntilExpiry, ocspError)
+		p.exportMetrics(peers[0].NotAfter, ocspError)
 		return false, time.Since(start)
 	}
-	if ocspStatus == ocsp.Unknown {
-		p.exportMetrics(timeUntilExpiry, ocspUnknown)
+
+	switch ocspStatus {
```

Comment:
I would invert this switch/case. Rather than switching on the received status, and then doing a check against `p.response`, switch on the value of `p.response` itself and then within each case check that the received status matches what we want.